jsk_robot_common/jsk_robot_startup/scripts/update_workspace_main.sh

```diff
@@ -94,12 +94,13 @@ if [ -n "$WSTOOL_STATUS" ]; then
     wstool diff -t $WORKSPACE/src | sed -e "s/'/ /g" -e "s/^--/ /g" -e 's/"/ /g' -e "s/<br>/\\\n/" -e 's/$/<br>/g' -e "s/,/ /g" | tee -a $TMP_MAIL_BODY_FILE
 fi
 if [ "${UPDATE_WORKSPACE}" == "true" ]; then
-    wstool foreach -t $WORKSPACE/src --git 'git stash'
-    wstool update -t $WORKSPACE/src jsk-ros-pkg/jsk_robot --delete-changed-uris
+    wstool foreach -t $WORKSPACE/src --git 'git stash -u'
+    wstool foreach -t $WORKSPACE/src --git 'git fetch origin --prune'
+    wstool update -t $WORKSPACE/src $(rospack find jsk_robot_startup)/../.. --delete-changed-uris
     ln -sf $ROSINSTALL $WORKSPACE/src/.rosinstall
     wstool update -t $WORKSPACE/src --delete-changed-uris
-    # Forcefully checkout specified branch
-    wstool foreach -t $WORKSPACE/src --git --shell 'branchname=$(git rev-parse --abbrev-ref HEAD); if [ $branchname != "HEAD" ]; then git reset --hard HEAD; git checkout origin/$branchname; git branch -D $branchname; git checkout -b $branchname --track origin/$branchname; fi'
+    # When the repository's Spec-Version branch has commits which are not pushed to remote, they are evacuated to another branch
+    wstool foreach -t $WORKSPACE/src --git --shell 'expected_version=$(wstool info . | grep Spec-Version: | awk '\''{print $2}'\''); current_version=$(git rev-parse --abbrev-ref HEAD); remote_diff=$(git diff --stat $current_version origin/$current_version); if [ "$expected_version" = "$current_version" ] && [ -n "$remote_diff" ]; then git checkout -b ${expected_version}-patch-$(date +%Y%m%d%H%M%S); git checkout origin/$expected_version; git branch -D $expected_version; git checkout -b $expected_version --track origin/$expected_version; fi'
```

Comment:
Why `remote_diff` should be checked? It looks like if there is no diff between remote and local but they are diverged, this scripts does not checkout remote branch.

Reply:
If you assume that there is still diff even after `git stash -u`, it seems required to commit that change to patched branch.

Reply:
When 

> there is no diff between remote and local but they are diverged

and 

> If you assume that there is still diff even after git stash -u

happens?

Reply:
It can happen you cherry pick some commit from a remote branch. What do you want to do with `remote_diff` ?

Reply:
It can also happen when squashing a branch to some point.

Reply:
And if you assume `there is no diff even after git stash -u`, `remote_diff` looks unnecessary.

Reply:
Previously, if you commit `develop/fetch` directly locally, the commits disappear because of `git branch -D` . I want to check remote_diff to find some differences between upstream and local, then evacuate commits. If no diff is found, which means `[ ! -n "$remote_diff" ]`, we do nothing.

Reply:
That feature is nice, I agree.

The flow of current code is

```
If there is uncommited change
    Stage them
Trying to the same name branch with remote target branch (we call this as current branch).
Delete original current branch
Checkout remote branch
```

And for my understanding, desired flow will be

```
If there is uncommited change
    Stage them
Trying to the same name branch to remote target branch (we call this as current branch).
If current branch is diverged from remote branch (= branch name is the same but commit ID is different)
    Evacuate current branch to renamed branch
    Delete original current branch
    Checkout remote target branch
```

And if you still assume there is uncommited changes, it will be

```
If there is uncommited change
    Stage them
Trying to the same name branch to remote target branch (we call this as current branch).
If current branch is diverged from remote branch (= branch name is the same but commit ID is different)
    commit all changes to branch
    Evacuate current branch to renamed branch
    Delete original current branch
    Checkout remote target branch
```

But your code looks like

```
If there is uncommited change
    Stage here
If current branch the same remote branch and no difference between them (!= no divergence)
    Evacuate current branch to renamed branch
    Delete original current branch
    Checkout remote branch
```